lib/ui/painting/path.cc

```diff
@@ -67,43 +67,92 @@ void CanvasPath::RegisterNatives(tonic::DartLibraryNatives* natives) {
                      FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
 }
 
-CanvasPath::CanvasPath() {}
+CanvasPath::CanvasPath()
+    : tracked_path_(std::make_shared<TrackedVolatilePath>()) {
+  resetVolatility();
+}
+
+CanvasPath::~CanvasPath() = default;
+
+void CanvasPath::resetVolatility() {
+  if (!tracked_path_->tracking_volatility) {
+    tracked_path_->path_.setIsVolatile(true);
+    tracked_path_->tracking_volatility = true;
+    {
+      std::scoped_lock guard(volatile_paths_mutex_);
+      volatile_paths_.insert(tracked_path_);
+    }
+  }
+}
 
-CanvasPath::~CanvasPath() {}
+// static
+std::mutex CanvasPath::volatile_paths_mutex_;
+// static
+std::unordered_set<std::shared_ptr<CanvasPath::TrackedVolatilePath>>
+    CanvasPath::volatile_paths_;
+// static
+void CanvasPath::updatePathVolatility() {
+  TRACE_EVENT0("flutter", "updatePathVolatility");
+  std::scoped_lock guard(volatile_paths_mutex_);
+  for (auto it = volatile_paths_.begin(), last = volatile_paths_.end();
+       it != last;) {
+    auto path = *it;
+    path->frame_count++;
+    if (path->frame_count >= number_of_frames_until_non_volatile) {
+      path->path_.setIsVolatile(false);
+      path->tracking_volatility = false;
+      it = volatile_paths_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+void CanvasPath::ReleaseDartWrappableReference() const {
+  std::scoped_lock guard(volatile_paths_mutex_);
+  volatile_paths_.erase(tracked_path_);
+}
 
 int CanvasPath::getFillType() {
-  return static_cast<int>(path_.getFillType());
+  return static_cast<int>(tracked_path_->path_.getFillType());
```

Comment:
Use `path()` to access the SkPath in these methods.  Add a private non-const mutable_path() accessor for methods that modify the path.